iwyu.cc

```diff
@@ -1375,62 +1375,70 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
+    SourceLocation spelling_loc = sm->getSpellingLoc(loc);
+
+    VERRS(5) << "Original use loc: " << PrintableLoc(use_loc) << "\n";
+    VERRS(5) << "Spelling loc: " << PrintableLoc(spelling_loc) << "\n";
+    VERRS(5) << "Expansion loc: " << PrintableLoc(expansion_loc) << "\n";
+
+    const FileEntry* macro_def_file = GetLocFileEntry(spelling_loc);
+    VERRS(5) << "Macro is defined in file '" << GetFilePath(macro_def_file)
+             << "'. Looking for fwd-decl hint...\n";
+
     const NamedDecl* fwd_decl = nullptr;
     for (const NamedDecl* redecl : GetClassRedecls(decl)) {
-      if (GetFileEntry(redecl) == macro_def_file &&
-          IsForwardDecl(redecl)) {
+      if (GetFileEntry(redecl) == macro_def_file && IsForwardDecl(redecl)) {
         fwd_decl = redecl;
+
+        // Make sure we keep that forward-declaration, even if it's probably
+        // unused in this file.
+        IwyuFileInfo* file_info =
+            preprocessor_info().FileInfoFor(macro_def_file);
+        file_info->ReportForwardDeclareUse(
+            spelling_loc, fwd_decl,
+            IsNodeInsideCXXMethodBody(current_ast_node()), nullptr);
         break;
       }
     }
 
-    if (fwd_decl) {
-      // Make sure we keep that forward-declaration, even if it's probably
-      // unused in this file.
-      preprocessor_info().FileInfoFor(macro_def_file)->ReportForwardDeclareUse(
-          use_loc, decl, IsNodeInsideCXXMethodBody(current_ast_node()),
-          nullptr);
-    }
-
-    // Resolve the best use location based on our current knowledge.
-    //
-    // 1) If the macro definition file forward-declares the used decl, that's a
-    //    hint that it wants the expansion location to take responsibility.
-    // 2) If the symbol being used is passed as an argument to the macro, we
-    //    want the expansion location to take responsibility (that thing they
-    //    passed in is something they know about, and the macro doesn't.)
-    // 3) If the use_loc is in <scratch space>, we assume it's formed by macro
-    //    argument concatenation, and attribute responsibility to the expansion
-    //    location.
-    //
-    // Otherwise, the macro file is responsible for including the symbol.
-    SourceLocation expansion_loc = GetInstantiationLoc(use_loc);
-    const char* responsible = "expansion";
-
     if (fwd_decl != nullptr) {
-      VERRS(5) << "Macro file has a forward-decl attribution hint.\n";
+      // If the macro file has a forward-decl, the expansion loc is responsible.
+      VERRS(5) << "Found a forward-decl in macro definition file.\n";
       use_loc = expansion_loc;
-    } else if (IsInScratchSpace(use_loc)) {
-      VERRS(5) << "Decl was used in <scratch space>, presumably as a result of "
-                  "macro arg concatenation.\n";
+    } else if (sm->isMacroArgExpansion(spelling_loc)) {
+      // For macro arguments, the expansion loc is responsible.
+      VERRS(5) << "Spelling location is in a macro-arg expansion.\n";
```

Comment:
Thanks, good feedback!

It sounds suspicious that this branch isn't hit by the test suite; I would have expected the new nested htons test to trigger this.

As for spelling vs expansion, my theory was that macro expansions are 'incremental'. So the 'unwind' at the top of the method would find the right level of expansion to make responsible for the use by repeatedly following immediate macro expansions.

Once we find the right point in the 'macro backtrace', the spelling loc should be that point and the expansion loc should be one level up. I've used debug traces of the macro backtrace to come to this conclusion. So `spelling_loc` and `expansion_loc` are relative, not the spelling/expansion locs you'd get if you asked the original use location.

I'm fully prepared to learn that this is all a huge misunderstanding :-)

I'll try to find some time to design a test case that hits the `isMacroArgExpansion` branch, or come up with an explanation for why it's not currently covered.